src/ethereum/vm/instructions.py

```diff
@@ -44,21 +68,313 @@ def add(evm: Evm) -> None:
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `3`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_MID`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_MID)
+
+    x, y, z = pop_multiple(evm.stack, 3)
+    res = x.unchecked_add(y) % z if z != 0 else U256(0)
+
+    push(evm.stack, res)
+
+
+def mulmod(evm: Evm) -> None:
+    """
+    Modulo multiplication of the top 2 elements with the 3rd element. Pushes
+    the result back on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `3`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_MID`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_MID)
+
+    x, y, z = pop_multiple(evm.stack, 3)
+    res = x.unchecked_mul(y) % z if z != 0 else U256(0)
+
+    push(evm.stack, res)
+
+
+def exp(evm: Evm) -> None:
+    """
+    Exponential operation of the top 2 elements. Pushes the result back on
+    the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_MID`.
+    """
+    x, y = pop_multiple(evm.stack, 2)
+
+    gas_used = GAS_EXP
+    if y != 0:
+        # This is equivalent to 1 + floor(log(y, 256)). But in python the log
+        # function is inaccurate leading to wrong results.
+        exponent_bits = y.bit_length()
+        exponent_bytes = (exponent_bits + 7) // 8
+        gas_used += GAS_EXP_BYTE * exponent_bytes
+    evm.gas_left = subtract_gas(evm.gas_left, gas_used)
+
+    res = cast(U256, pow(x, y, U256_MAX_VALUE + 1))
+
+    push(evm.stack, res)
+
+
+def signextend(evm: Evm) -> None:
+    """
+    Sign Extend operation. In other words, we extend a signed number which
+    fits in N bytes to 32 bytes.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_LOW)
+
+    # byte_num would be 0-indexed when inserted to the stack.
+    byte_num, value = pop_multiple(evm.stack, 2)
+
+    if byte_num > 31:
+        # Can't extend any further
+        res = value
+    else:
+        # U256(0).to_be_bytes() gives b'' instead b'\x00'. # noqa: SC100
+        value_bytes = value.to_be_bytes() or b"\x00"
+
+        # Now among the obtained value bytes, consider only
+        # N `least significant bytes`, where N is `byte_num + 1`.
+        value_bytes = value_bytes[len(value_bytes) - 1 - byte_num :]
```

Comment:
I _think_ this can be simplified to `value_bytes = value_bytes[-1 - byte_num :]`.

Reply:
I think avoiding fun python indexing is probably preferable.

Reply:
I assume we are ok with negative indexing. Going ahead with it.